rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -68,16 +92,25 @@ class MessagePoolMemoryStrategy
    */
   std::shared_ptr<MessageT> borrow_message()
   {
-    size_t current_index = next_array_index_;
-    next_array_index_ = (next_array_index_ + 1) % Size;
-    if (pool_[current_index].used) {
-      throw std::runtime_error("Tried to access message that was still in use! Abort.");
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+    if (free_list_.size() == 0) {
+      throw std::runtime_error("No more free slots in the pool");
     }
-    pool_[current_index].msg_ptr_->~MessageT();
-    new (pool_[current_index].msg_ptr_.get())MessageT;
 
-    pool_[current_index].used = true;
-    return pool_[current_index].msg_ptr_;
+    size_t current_index = free_list_.pop_front();
+
+    return std::shared_ptr<MessageT>(
+      new(pool_[current_index]) MessageT(),
+      [this](MessageT * p) {
+        std::lock_guard<std::mutex> lock(pool_mutex_);
+        for (size_t i = 0; i < Size; ++i) {
+          if (pool_[i] == p) {
+            p->~MessageT();
+            free_list_.push_back(i);
+            break;
+          }
+        }
+      });
```

Comment:
If the pool is destroyed before the borrowed shared pointers, then this code is broken too.

So either we need to just say this is undefined behavior (destroying the pool before all borrows are returned) or we need to improve this code.

If we do the first, then your warning about leaking memory is not necessary anymore (imo) since it's broken no matter what.

To fix this code, you'd need to make this pool an `enable_shared_from_this` and then capture a weak_ptr to it in this lambda, to avoid a circular reference. And then this code would need to just run the destructor and not check the pool or free list if the weak_ptr was unable to `lock()`.

Or alternatively you could capture the `pool_` itself (and the `free_list_`) and make it a shared pointer, and move the free'ing logic into its destructor.

Reply:
> Or alternatively you could capture the `pool_` itself (and the `free_list_`) and make it a shared pointer, and move the free'ing logic into its destructor.

Yeah, this is all a really good point.  I ended up going this direction, where all of the member variables are `shared_ptr`, and hence can outlive this class.  See ec86e41 .  I verified with this code that holding the references past when `MessagePoolMemoryStrategy` is destroyed does the right thing (at least, valgrind gave me no errors in that situation).

Please take another look.